File: model/src/main/java/io/github/chriszhong/model/organization/event/RequiresEventFactory.java
package io.github.chriszhong.model.organization.event;

import io.github.runtimemodels.chazm.relation.Requires;

/**
 * The {@linkplain RequiresEventFactory} interface defines the API for constructing {@linkplain RequiresEvent}s.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
@FunctionalInterface
public interface RequiresEventFactory {

	/**
	 * Constructs an {@linkplain RequiresEvent}.
	 *
	 * @param category
	 *            the {@linkplain EventCategory}.
	 * @param requires
	 *            the {@linkplain Requires}.
	 *
	 * @return a {@linkplain RequiresEvent}.
	 */
	RequiresEvent build(EventCategory category, Requires requires);

}

File: chazm-model/src/main/kotlin/runtimemodels/chazm/model/relation/RequiresEventFactory.kt
package runtimemodels.chazm.model.relation

import runtimemodels.chazm.api.relation.Requires
import runtimemodels.chazm.model.event.EventType

/**
 * The [RequiresEventFactory] interface defines the API for constructing [RequiresEvent]s.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
@FunctionalInterface
interface RequiresEventFactory {
    /**
     * Constructs an [RequiresEvent].
     *
     * @param category the [EventType].
     * @param requires the [Requires].
     * @return a [RequiresEvent].
     */
    fun build(category: EventType, requires: Requires): RequiresEvent
}
